// react-native-quiz-maker/lib/commonjs/components/question-types/Matching/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = require("../../../styles");

var _texts = require("../../texts");

var _PlayingCard = _interopRequireDefault(require("./PlayingCard"));

var _setup = require("./setup");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const MatchingQuestion = /*#__PURE__*/_react.default.memo(({
  questionAnswerPairs,
  onSubmit = isCorrect => isCorrect,
  customContainerStyle,
  instructionText,
  instructionsTextStyle,
  questionTextStyle,
  headerContainerStyle,
  cardListContainerStyle,
  correctCardColor,
  incorrectCardColor,
  inactiveCardColor,
  cardStyle,
  activeCardColor,
  cardTextStyle,
  cardListStyle,
  listScrollEnabled
}) => {
  const initialDeck = (0, _setup.setup)(questionAnswerPairs);
  const [state, dispatch] = (0, _react.useReducer)(_setup.matchingReducer, {
    deck: initialDeck,
    continueEnabled: false
  });

  const selectCard = selection => dispatch({
    type: 'selectCard',
    payload: {
      selection,
      submit: onSubmit
    }
  });

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, customContainerStyle]
  }, /*#__PURE__*/_react.default.createElement(_texts.QuestionHeader, {
    instructions: instructionText || 'Please insert instructions here',
    instructionsTextStyle: instructionsTextStyle,
    questionTextStyle: questionTextStyle,
    headerContainerStyle: headerContainerStyle
  }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.cardListContainer, cardListContainerStyle]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.FlatList, {
    data: state.deck,
    numColumns: 2,
    keyExtractor: item => item.name,
    renderItem: ({
      item
    }) => /*#__PURE__*/_react.default.createElement(_PlayingCard.default, {
      item: item,
      selectCard: selectCard,
      correctCardColor: correctCardColor,
      incorrectCardColor: incorrectCardColor,
      cardStyle: cardStyle,
      inactiveCardColor: inactiveCardColor,
      activeCardColor: activeCardColor,
      cardTextStyle: cardTextStyle
    }),
    style: [cardListStyle],
    scrollEnabled: listScrollEnabled || false
  })));
});

const styles = _reactNative.StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    width: _styles.Sizing.sw
  },
  cardListContainer: {
    flex: 4
  }
});

var _default = MatchingQuestion;
exports.default = _default;
