// react-native-quiz-maker/lib/commonjs/components/question-types/Matching/PlayingCard.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _services = require("../../../services");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const appropriateCardOpacity = (visible, justSubmittedAndIsCorrect) => {
  if (justSubmittedAndIsCorrect) {
    return new _reactNative.Animated.Value(1);
  } else if (!visible) {
    return new _reactNative.Animated.Value(0);
  } else {
    return new _reactNative.Animated.Value(1);
  }
};

const PlayingCard = ({
  item,
  selectCard,
  cardStyle = {},
  correctCardColor = '#73D413',
  incorrectCardColor = '#F33232',
  inactiveCardColor = 'white',
  activeCardColor = '#4294DB',
  cardTextStyle
}) => {
  const {
    selected,
    visible,
    name,
    justSubmitted
  } = item;
  var disabled = !visible;
  const justSubmittedAndIsCorrect = justSubmitted && !visible;
  const justSubmittedAndIsIncorrect = justSubmitted && visible;
  const cardOpacity = (0, _react.useState)(appropriateCardOpacity(visible, justSubmittedAndIsCorrect))[0];
  const animatedValue = new _reactNative.Animated.Value(90);

  const fadeCardOut = () => (0, _services.timedAnimation)(cardOpacity, 800, 0).start();

  const interpolateToWhite = () => {
    _reactNative.Animated.timing(animatedValue, {
      toValue: 255,
      duration: 700,
      useNativeDriver: false
    }).start();
  };

  const interpolateColor = animatedValue.interpolate({
    inputRange: [0, 250],
    outputRange: [incorrectCardColor, 'rgb(250, 250, 250)']
  });
  const bgColor = {
    backgroundColor: appropriateCardColor()
  };

  function appropriateCardColor() {
    if (justSubmittedAndIsCorrect) {
      return correctCardColor;
    } else if (justSubmittedAndIsIncorrect) {
      return interpolateColor;
    } else if (selected) {
      return activeCardColor;
    } else {
      return inactiveCardColor;
    }
  }

  (0, _react.useEffect)(() => {
    if (justSubmittedAndIsCorrect) {
      fadeCardOut();
    } else if (justSubmittedAndIsIncorrect) {
      interpolateToWhite();
    }
  });
  return /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    style: [styles.PlayingCard, cardStyle],
    disabled: disabled,
    onPress: () => selectCard(item)
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    style: [{
      opacity: cardOpacity,
      width: '100%',
      height: '100%',
      borderRadius: _reactNative.Dimensions.get('screen').height / 18,
      justifyContent: 'center'
    }, bgColor]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: [styles.text, cardTextStyle]
  }, name)));
};

const styles = _reactNative.StyleSheet.create({
  PlayingCard: {
    height: _reactNative.Dimensions.get('screen').height / 12,
    width: _reactNative.Dimensions.get('screen').width / 3.4,
    borderWidth: 2,
    borderColor: 'grey',
    borderRadius: _reactNative.Dimensions.get('screen').height / 18,
    margin: 5,
    justifyContent: 'center',
    alignItems: 'center'
  },
  text: {
    alignSelf: 'center'
  }
});

var _default = PlayingCard;
exports.default = _default;
