// react-native-quiz-maker/lib/commonjs/components/question-types/Matching/setup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.matchingReducer = exports.setup = void 0;

const compose = (...fns) => x => fns.reduceRight((y, f) => f(y), x);

const attemptToMatch = (card, deck) => compose(updateDeckAfterSubmission, checkForMatch, selectCard)({
  card,
  deck
});

const checkForMatch = deck => {
  //returns true or false based on 2 selected cards
  var selectedCards = deck.filter(c => c.selected);

  if (selectedCards.length == 2) {
    return {
      submission: selectedCards[0].pair == selectedCards[1].pair,
      deck
    };
  } else {
    return {
      submission: false,
      deck
    };
  }
};

const updateDeckAfterSubmission = verdict => {
  //returns new array - if selected card are correct then they become invisible, if not then they become unselected
  return verdict.submission ? verdict.deck.map(c => {
    return c.selected ? { ...c,
      selected: false,
      visible: false,
      justSubmitted: true
    } : { ...c,
      justSubmitted: false
    };
  }) : verdict.deck.map(c => {
    return c.selected ? { ...c,
      selected: false,
      justSubmitted: true
    } : { ...c,
      justSubmitted: false
    };
  });
};

const selectCard = obj => {
  //sets selected card.selected = true and returns array of updated cards
  return obj.deck.map(c => {
    return c.index == obj.card.index ? { ...c,
      selected: true,
      justSubmitted: false
    } : { ...c,
      justSubmitted: false
    };
  });
};

const unselectCard = obj => {
  return obj.deck.map(c => {
    return c.index == obj.card.index ? { ...c,
      selected: false,
      justSubmitted: false
    } : { ...c,
      justSubmitted: false
    };
  });
};

const splitQuestionAnswerPair = (questionAnswer, increment, index) => {
  return [{
    index: index,
    name: questionAnswer.question,
    pair: index,
    selected: false,
    visible: true,
    justSubmitted: false
  }, {
    index: increment + index,
    name: questionAnswer.answer,
    pair: index,
    selected: false,
    visible: true,
    justSubmitted: false
  }];
};

const createMatchingCards = questionAnswers => {
  const incrementValue = questionAnswers.length;
  const matchingCards = questionAnswers.flatMap((curr, index) => splitQuestionAnswerPair(curr, incrementValue, index));
  return matchingCards;
};

const shuffleArray = ogArray => {
  let array = [...ogArray];

  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }

  return array;
};

const setup = compose(shuffleArray, createMatchingCards);
exports.setup = setup;

const matchingReducer = (state, action) => {
  if (action.type === 'selectCard') {
    var isUserPickingSecondCard = state.deck.filter(c => c.selected).length > 0;
    var isUserPickingSameCard = state.deck.find(c => c.selected) === action.payload.selection;
    var isGameFinished = state.deck.filter(c => c.visible).length <= 2 && isUserPickingSecondCard;
    isGameFinished && action.payload.submit(true);
    return isUserPickingSecondCard ? isUserPickingSameCard ? { ...state,
      deck: unselectCard({
        card: action.payload.selection,
        deck: state.deck
      })
    } : { ...state,
      deck: attemptToMatch(action.payload.selection, state.deck),
      continueEnabled: isGameFinished
    } : //setAllCards(compose( updateDeckAfterSubmission, checkForMatch, selectCard )({card, allCards}))
    { ...state,
      deck: selectCard({
        card: action.payload.selection,
        deck: state.deck
      })
    }; //setAllCards(selectCard({card, allCards}))
  } else if (action.type === 'reset') {
    return {
      deck: setup(action.payload.questionAnswerPairs),
      continueEnabled: false
    };
  } else {
    return { ...state
    };
  }
};

exports.matchingReducer = matchingReducer;
