// react-native-quiz-maker/lib/module/components/question-types/Writing/index.js
import React, { useReducer } from 'react';
import { View, TextInput } from 'react-native';
import { Sizing } from '../../../styles';
import { GiveUpButton } from '../../buttons';
import { QuestionHeader } from '../../texts';
import styles from '../styles';
import { WritingQInitialState } from './definitions';
import { WritingQReducer } from './reducer';

const WritingQuestion = ({
  onSubmit = isCorrect => isCorrect,
  answer,
  instructionText,
  question,
  giveUpButtonContainerStyle,
  giveUpButtonStyle,
  giveUpLabelStyle
}) => {
  const [state, dispatch] = useReducer(WritingQReducer, WritingQInitialState);

  const handleResponse = response => {
    if (response == answer) {
      dispatch({
        type: 'correctAnswer',
        response
      });
      onSubmit(true);
    } else {
      dispatch({
        type: 'handleResponse',
        response
      });
    }
  };

  const giveUp = () => {
    dispatch({
      type: 'giveUp',
      payload: {
        answer: answer
      }
    });
    onSubmit(false);
  };

  return /*#__PURE__*/React.createElement(View, {
    style: {
      width: Sizing.sw
    }
  }, /*#__PURE__*/React.createElement(QuestionHeader, {
    instructions: instructionText,
    question: question
  }), /*#__PURE__*/React.createElement(View, {
    style: {
      flex: 3,
      justifyContent: 'center'
    }
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.underline
  }, /*#__PURE__*/React.createElement(TextInput, {
    onChangeText: response => handleResponse(response),
    value: state.inputValue.toString(),
    editable: state.inputEnabled,
    autoCorrect: false,
    autoCapitalize: "sentences",
    placeholder: "start typing here...",
    style: {
      width: Sizing.sw / 2,
      textAlign: 'center'
    }
  }))), /*#__PURE__*/React.createElement(GiveUpButton, {
    onGiveUp: giveUp,
    labelStyle: giveUpLabelStyle,
    buttonStyle: giveUpButtonStyle,
    buttonContainerStyle: giveUpButtonContainerStyle,
    enabled: state.inputEnabled
  }));
};

WritingQuestion.defaultProps = {
  question: 'Insert question here?',
  answer: 'answer'
};
export default WritingQuestion;
