// react-native-quiz-maker/lib/module/components/question-types/Matching/index.js
import React, { useReducer } from 'react';
import { FlatList, View, StyleSheet } from 'react-native';
import { Sizing } from '../../../styles';
import { QuestionHeader } from '../../texts';
import PlayingCard from './PlayingCard';
import { matchingReducer, setup } from './setup';
const MatchingQuestion = /*#__PURE__*/React.memo(({
  questionAnswerPairs,
  onSubmit = isCorrect => isCorrect,
  customContainerStyle,
  instructionText,
  instructionsTextStyle,
  questionTextStyle,
  headerContainerStyle,
  cardListContainerStyle,
  correctCardColor,
  incorrectCardColor,
  inactiveCardColor,
  cardStyle,
  activeCardColor,
  cardTextStyle,
  cardListStyle,
  listScrollEnabled
}) => {
  const initialDeck = setup(questionAnswerPairs);
  const [state, dispatch] = useReducer(matchingReducer, {
    deck: initialDeck,
    continueEnabled: false
  });

  const selectCard = selection => dispatch({
    type: 'selectCard',
    payload: {
      selection,
      submit: onSubmit
    }
  });

  return /*#__PURE__*/React.createElement(View, {
    style: [styles.container, customContainerStyle]
  }, /*#__PURE__*/React.createElement(QuestionHeader, {
    instructions: instructionText || 'Please insert instructions here',
    instructionsTextStyle: instructionsTextStyle,
    questionTextStyle: questionTextStyle,
    headerContainerStyle: headerContainerStyle
  }), /*#__PURE__*/React.createElement(View, {
    style: [styles.cardListContainer, cardListContainerStyle]
  }, /*#__PURE__*/React.createElement(FlatList, {
    data: state.deck,
    numColumns: 2,
    keyExtractor: item => item.name,
    renderItem: ({
      item
    }) => /*#__PURE__*/React.createElement(PlayingCard, {
      item: item,
      selectCard: selectCard,
      correctCardColor: correctCardColor,
      incorrectCardColor: incorrectCardColor,
      cardStyle: cardStyle,
      inactiveCardColor: inactiveCardColor,
      activeCardColor: activeCardColor,
      cardTextStyle: cardTextStyle
    }),
    style: [cardListStyle],
    scrollEnabled: listScrollEnabled || false
  })));
});
const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    width: Sizing.sw
  },
  cardListContainer: {
    flex: 4
  }
});
export default MatchingQuestion;
