// react-native-quiz-maker/lib/commonjs/components/buttons/Draggable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativeGestureHandler = require("react-native-gesture-handler");

var _timedAnimation = _interopRequireDefault(require("../../services/timedAnimation"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const onPanGestureEvent = _translate => {
  return _reactNative.Animated.event([{
    nativeEvent: {
      translationX: _translate.x,
      translationY: _translate.y
    }
  }], {
    useNativeDriver: true
  });
};

const DraggableButton = props => {
  const _translate = (0, _react.useState)(new _reactNative.Animated.ValueXY({
    x: 0,
    y: 0
  }))[0];
  const [layout, setLayout] = (0, _react.useState)({
    x: 0,
    y: 0
  });

  const _onPanGestureEvent = onPanGestureEvent(_translate);

  const _onHandlerStateChange = props.onHandlerStateChange(layout, _translate, props.index);

  (0, _react.useEffect)(() => {
    if (props.landingZoneOccupier == props.index) {
      (0, _timedAnimation.default)(_translate, 0, {
        x: 0,
        y: 0
      }).start(() => {
        _translate.setOffset({
          x: 0,
          y: 0
        });

        _translate.setValue({
          x: 0,
          y: 0
        });

        props.clearOut();
      });
    }
  }, [props.choices]);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.main,
    onLayout: event => {
      setLayout({
        x: event.nativeEvent.layout.x,
        y: event.nativeEvent.layout.y
      });
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNativeGestureHandler.PanGestureHandler, {
    onGestureEvent: _onPanGestureEvent,
    onHandlerStateChange: _onHandlerStateChange,
    hitSlop: {
      horizontal: 10,
      vertical: 10
    },
    enabled: props.enabled
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, null, /*#__PURE__*/_react.default.createElement(_reactNativeGestureHandler.TapGestureHandler, {
    onHandlerStateChange: _onHandlerStateChange,
    enabled: props.enabled
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    style: [{
      transform: [{
        translateX: _translate.x
      }, {
        translateY: _translate.y
      }]
    }]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      width: 100,
      ...styles.choice
    },
    key: props.index
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.text
  }, props.name)))))));
};

const styles = _reactNative.StyleSheet.create({
  main: {
    alignSelf: 'center',
    marginVertical: 10,
    backgroundColor: '#f2f2f2',
    borderRadius: 23
  },
  choice: {
    height: 46,
    backgroundColor: 'white',
    borderWidth: 2,
    borderRadius: 23,
    justifyContent: 'center',
    shadowOpacity: 0.25,
    shadowRadius: 2,
    shadowColor: 'rgb(0, 0, 0)',
    shadowOffset: {
      height: 1,
      width: 0
    }
  },
  text: {
    fontSize: 14,
    textAlign: 'center'
  }
});

var _default = DraggableButton;
exports.default = _default;
