// react-native-quiz-maker/lib/commonjs/components/buttons/Primary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const PrimaryButton = props => /*#__PURE__*/React.createElement(_reactNative.SafeAreaView, {
  style: [styles.buttonContainer, props.buttonContainerStyle]
}, /*#__PURE__*/React.createElement(_reactNative.TouchableHighlight, {
  onPress: () => props.onPress(),
  disabled: !props.enabled,
  style: [styles.button, props.buttonStyle]
}, /*#__PURE__*/React.createElement(_reactNative.Text, {
  style: [styles.label, props.labelStyle]
}, props.label)));

const styles = _reactNative.StyleSheet.create({
  buttonContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  button: {
    backgroundColor: 'blue',
    width: _reactNative.Dimensions.get('screen').width / 3,
    height: _reactNative.Dimensions.get('screen').width / 10,
    justifyContent: 'center',
    borderRadius: 10
  },
  label: {
    color: 'white',
    textAlign: 'center'
  }
});

var _default = PrimaryButton;
exports.default = _default;
