// react-native-quiz-maker/lib/commonjs/components/question-types/Writing/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = require("../../../styles");

var _buttons = require("../../buttons");

var _texts = require("../../texts");

var _styles2 = _interopRequireDefault(require("../styles"));

var _definitions = require("./definitions");

var _reducer = require("./reducer");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const WritingQuestion = ({
  onSubmit = isCorrect => isCorrect,
  answer,
  instructionText,
  question,
  giveUpButtonContainerStyle,
  giveUpButtonStyle,
  giveUpLabelStyle
}) => {
  const [state, dispatch] = (0, _react.useReducer)(_reducer.WritingQReducer, _definitions.WritingQInitialState);

  const handleResponse = response => {
    if (response == answer) {
      dispatch({
        type: 'correctAnswer',
        response
      });
      onSubmit(true);
    } else {
      dispatch({
        type: 'handleResponse',
        response
      });
    }
  };

  const giveUp = () => {
    dispatch({
      type: 'giveUp',
      payload: {
        answer: answer
      }
    });
    onSubmit(false);
  };

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      width: _styles.Sizing.sw
    }
  }, /*#__PURE__*/_react.default.createElement(_texts.QuestionHeader, {
    instructions: instructionText,
    question: question
  }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      flex: 3,
      justifyContent: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: _styles2.default.underline
  }, /*#__PURE__*/_react.default.createElement(_reactNative.TextInput, {
    onChangeText: response => handleResponse(response),
    value: state.inputValue.toString(),
    editable: state.inputEnabled,
    autoCorrect: false,
    autoCapitalize: "sentences",
    placeholder: "start typing here...",
    style: {
      width: _styles.Sizing.sw / 2,
      textAlign: 'center'
    }
  }))), /*#__PURE__*/_react.default.createElement(_buttons.GiveUpButton, {
    onGiveUp: giveUp,
    labelStyle: giveUpLabelStyle,
    buttonStyle: giveUpButtonStyle,
    buttonContainerStyle: giveUpButtonContainerStyle,
    enabled: state.inputEnabled
  }));
};

WritingQuestion.defaultProps = {
  question: 'Insert question here?',
  answer: 'answer'
};
var _default = WritingQuestion;
exports.default = _default;
