// react-native-quiz-maker/lib/module/components/question-types/MultipleChoice/index.js
import React, { useReducer } from 'react';
import { View } from 'react-native';
import { State } from 'react-native-gesture-handler';
import { DraggableButton, PrimaryButton } from '../../buttons';
import { QuestionHeader } from '../../texts';
import timedAnimation from '../../../services/timedAnimation';
import { MultipleChoiceQInitialState } from './definitions';
import { MultipleChoiceQReducer } from './reducer';
import { Sizing } from '../../../styles';
import styles from '../styles';

const MultipleChoiceQuestion = ({
  answer,
  allChoices,
  onSubmit = isCorrect => isCorrect,
  customContainerStyle,
  instructionText,
  question,
  checkButtonContainerStyle,
  checkLabelStyle,
  checkButtonStyle
}) => {
  const [state, dispatch] = useReducer(MultipleChoiceQReducer, MultipleChoiceQInitialState);

  const setLandingZoneCoordinates = layout => {
    dispatch({
      type: 'setCoordinates',
      payload: {
        layout
      }
    });
  };

  const prepareForLanding = occupier => dispatch({
    type: 'prepareForLanding',
    payload: {
      occupier
    }
  });

  const clearOut = () => dispatch({
    type: 'clearOut'
  });

  const selectChoice = selectedChoice => {
    selectedChoice == -1 ? dispatch({
      type: 'disableCheck'
    }) : dispatch({
      type: 'selectChoice',
      payload: {
        selectedChoice
      }
    });
  };

  const submit = () => {
    dispatch({
      type: 'checkPlease',
      payload: {
        answer: answer,
        allChoices: allChoices
      }
    });
    onSubmit(allChoices[state.selectedChoice] == answer);
  };

  const returnHandlerHome = (_translate, index) => {
    timedAnimation(_translate, 200, {
      x: 0,
      y: 0
    }).start(() => {
      _translate.setOffset({
        x: 0,
        y: 0
      });

      _translate.setValue({
        x: 0,
        y: 0
      });

      state.occupier == index && clearOut();
    });
  };

  const sendHandlerToPlace = (_layout, _translate, index) => {
    timedAnimation(_translate, 200, {
      x: 0,
      y: -state.layout.y - _layout.y + 50
    }).start(() => {
      _translate.setOffset({
        x: 0,
        y: -state.layout.y - _layout.y + 50
      });

      _translate.setValue({
        x: 0,
        y: 0
      });

      prepareForLanding(index);
      selectChoice(index);
    });
  };

  const onHandlerStateChange = (_layout, _translate, index) => event => {
    if (event.nativeEvent.oldState === State.ACTIVE) {
      _translate.flattenOffset();

      state.isFull ? returnHandlerHome(_translate, index) : sendHandlerToPlace(_layout, _translate, index);
    }
  };

  return /*#__PURE__*/React.createElement(View, {
    style: [{
      width: Sizing.sw
    }, customContainerStyle]
  }, /*#__PURE__*/React.createElement(QuestionHeader, {
    instructions: instructionText,
    question: question
  }), /*#__PURE__*/React.createElement(View, {
    onLayout: event => {
      setLandingZoneCoordinates({
        x: event.nativeEvent.layout.x,
        y: event.nativeEvent.layout.y
      });
    },
    style: {
      justifyContent: 'flex-start',
      alignItems: 'center',
      flex: 2
    }
  }, /*#__PURE__*/React.createElement(View, {
    style: { ...styles.underline,
      width: Sizing.normalize(answer.length * 20)
    }
  })), /*#__PURE__*/React.createElement(View, {
    style: {
      flex: 3
    }
  }, allChoices.map((name, index) => /*#__PURE__*/React.createElement(DraggableButton, {
    key: index,
    name: name.toString(),
    onHandlerStateChange: onHandlerStateChange,
    index: index,
    landingZoneOccupier: state.occupier,
    clearOut: clearOut,
    choices: allChoices,
    enabled: state.choicesEnabled
  }))), /*#__PURE__*/React.createElement(PrimaryButton, {
    label: "Check",
    onPress: submit,
    labelStyle: checkLabelStyle,
    buttonStyle: checkButtonStyle,
    buttonContainerStyle: checkButtonContainerStyle,
    enabled: state.checkEnabled
  }));
};

export default MultipleChoiceQuestion;
