// react-native-quiz-maker/lib/commonjs/components/question-types/MultipleChoice/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativeGestureHandler = require("react-native-gesture-handler");

var _buttons = require("../../buttons");

var _texts = require("../../texts");

var _timedAnimation = _interopRequireDefault(require("../../../services/timedAnimation"));

var _definitions = require("./definitions");

var _reducer = require("./reducer");

var _styles = require("../../../styles");

var _styles2 = _interopRequireDefault(require("../styles"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const MultipleChoiceQuestion = ({
  answer,
  allChoices,
  onSubmit = isCorrect => isCorrect,
  customContainerStyle,
  instructionText,
  question,
  checkButtonContainerStyle,
  checkLabelStyle,
  checkButtonStyle
}) => {
  const [state, dispatch] = (0, _react.useReducer)(_reducer.MultipleChoiceQReducer, _definitions.MultipleChoiceQInitialState);

  const setLandingZoneCoordinates = layout => {
    dispatch({
      type: 'setCoordinates',
      payload: {
        layout
      }
    });
  };

  const prepareForLanding = occupier => dispatch({
    type: 'prepareForLanding',
    payload: {
      occupier
    }
  });

  const clearOut = () => dispatch({
    type: 'clearOut'
  });

  const selectChoice = selectedChoice => {
    selectedChoice == -1 ? dispatch({
      type: 'disableCheck'
    }) : dispatch({
      type: 'selectChoice',
      payload: {
        selectedChoice
      }
    });
  };

  const submit = () => {
    dispatch({
      type: 'checkPlease',
      payload: {
        answer: answer,
        allChoices: allChoices
      }
    });
    onSubmit(allChoices[state.selectedChoice] == answer);
  };

  const returnHandlerHome = (_translate, index) => {
    (0, _timedAnimation.default)(_translate, 200, {
      x: 0,
      y: 0
    }).start(() => {
      _translate.setOffset({
        x: 0,
        y: 0
      });

      _translate.setValue({
        x: 0,
        y: 0
      });

      state.occupier == index && clearOut();
    });
  };

  const sendHandlerToPlace = (_layout, _translate, index) => {
    (0, _timedAnimation.default)(_translate, 200, {
      x: 0,
      y: -state.layout.y - _layout.y + 50
    }).start(() => {
      _translate.setOffset({
        x: 0,
        y: -state.layout.y - _layout.y + 50
      });

      _translate.setValue({
        x: 0,
        y: 0
      });

      prepareForLanding(index);
      selectChoice(index);
    });
  };

  const onHandlerStateChange = (_layout, _translate, index) => event => {
    if (event.nativeEvent.oldState === _reactNativeGestureHandler.State.ACTIVE) {
      _translate.flattenOffset();

      state.isFull ? returnHandlerHome(_translate, index) : sendHandlerToPlace(_layout, _translate, index);
    }
  };

  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [{
      width: _styles.Sizing.sw
    }, customContainerStyle]
  }, /*#__PURE__*/_react.default.createElement(_texts.QuestionHeader, {
    instructions: instructionText,
    question: question
  }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    onLayout: event => {
      setLandingZoneCoordinates({
        x: event.nativeEvent.layout.x,
        y: event.nativeEvent.layout.y
      });
    },
    style: {
      justifyContent: 'flex-start',
      alignItems: 'center',
      flex: 2
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: { ..._styles2.default.underline,
      width: _styles.Sizing.normalize(answer.length * 20)
    }
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      flex: 3
    }
  }, allChoices.map((name, index) => /*#__PURE__*/_react.default.createElement(_buttons.DraggableButton, {
    key: index,
    name: name.toString(),
    onHandlerStateChange: onHandlerStateChange,
    index: index,
    landingZoneOccupier: state.occupier,
    clearOut: clearOut,
    choices: allChoices,
    enabled: state.choicesEnabled
  }))), /*#__PURE__*/_react.default.createElement(_buttons.PrimaryButton, {
    label: "Check",
    onPress: submit,
    labelStyle: checkLabelStyle,
    buttonStyle: checkButtonStyle,
    buttonContainerStyle: checkButtonContainerStyle,
    enabled: state.checkEnabled
  }));
};

var _default = MultipleChoiceQuestion;
exports.default = _default;
