// react-native-quiz-maker/lib/commonjs/components/question-types/MultipleChoice/reducer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultipleChoiceQReducer = void 0;

var _definitions = require("./definitions");

const MultipleChoiceQReducer = (state, action) => {
  if (action.type === 'selectChoice') {
    return { ...state,
      selectedChoice: action.payload.selectedChoice,
      checkEnabled: true
    };
  } else if (action.type === 'checkPlease') {
    return action.payload.allChoices[state.selectedChoice] === action.payload.answer ? { ...state,
      questionStatus: 'correct',
      choicesEnabled: false,
      continueEnabled: true,
      checkEnabled: false
    } : { ...state,
      questionStatus: 'incorrect',
      choicesEnabled: false,
      continueEnabled: true,
      checkEnabled: false
    };
  } else if (action.type === 'disableCheck') {
    return { ...state,
      checkEnabled: false
    };
  } else if (action.type === 'setCoordinates') {
    return { ...state,
      layout: action.payload.layout
    };
  } else if (action.type === 'prepareForLanding') {
    return { ...state,
      isFull: true,
      occupier: action.payload.occupier
    };
  } else if (action.type === 'clearOut') {
    return { ...state,
      isFull: false,
      occupier: -1,
      checkEnabled: false
    };
  } else if (action.type === 'reset') {
    return { ..._definitions.MultipleChoiceQInitialState
    };
  } else {
    return state;
  }
};

exports.MultipleChoiceQReducer = MultipleChoiceQReducer;
