// react-native-quiz-maker/lib/commonjs/components/quiz-container/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _styles = require("../../styles");

var _services = require("../../services");

var _Matching = _interopRequireDefault(require("../question-types/Matching"));

var _MultipleChoice = _interopRequireDefault(require("../question-types/MultipleChoice"));

var _Writing = _interopRequireDefault(require("../question-types/Writing"));

var _reducer = require("./reducer");

var _buttons = require("../buttons");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const QuizContainer = ({
  questions,
  customContainerStyle,
  lives = 0,
  continueLabelStyle,
  continueButtonStyle,
  onSubmit,
  onComplete
}) => {
  const [state, dispatch] = (0, _react.useReducer)(_reducer.QuizContainerReducer, {
    questionComponents: [...questions],
    activeQuestion: 0,
    progress: 0,
    lives,
    continueEnabled: false
  });
  const didMount = (0, _react.useRef)(false);
  const horizontalContainer = (0, _react.useState)(new _reactNative.Animated.Value(0))[0];
  const progressIncrement = 100 / questions.length;

  const nextQuestion = () => state.activeQuestion === state.questionComponents.length - 1 ? onComplete(state.progress) : dispatch({
    type: 'nextSlide'
  });

  const userSubmit = (isCorrect = true) => {
    dispatch({
      type: 'userSubmit',
      payload: {
        isCorrect,
        progressIncrement
      }
    });
    onSubmit(isCorrect);
  };

  (0, _react.useEffect)(() => {
    didMount.current ? (0, _services.springAnimation)(horizontalContainer, state.activeQuestion).start() : didMount.current = true;
  }, [state.activeQuestion, horizontalContainer]);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [{
      flex: 1
    }, customContainerStyle]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Animated.View, {
    style: [styles.animatedContainer, {
      transform: [{
        translateX: horizontalContainer.interpolate({
          inputRange: [0, state.questionComponents.length],
          outputRange: [_styles.Sizing.sw * state.questionComponents.length / 2 - _styles.Sizing.sw / 2, -_styles.Sizing.sw * state.questionComponents.length / 2 - _styles.Sizing.sw / 2]
        })
      }]
    }]
  }, state.questionComponents.map((question, index) => {
    if (question.questionType === 'MultipleChoice' && 'question' in question && 'allChoices' in question) {
      return /*#__PURE__*/_react.default.createElement(_MultipleChoice.default, {
        key: index,
        question: question.question,
        answer: question.answer,
        allChoices: question.allChoices,
        onSubmit: userSubmit,
        instructionText: question.instructionText
      });
    } else if (question.questionType === 'Matching' && 'questionAnswerPairs' in question) {
      return /*#__PURE__*/_react.default.createElement(_Matching.default, {
        key: index,
        questionAnswerPairs: question.questionAnswerPairs,
        onSubmit: userSubmit,
        instructionText: question.instructionText
      });
    } else if (question.questionType === 'Writing' && 'question' in question) {
      return /*#__PURE__*/_react.default.createElement(_Writing.default, {
        key: index,
        question: question.question,
        answer: question.answer,
        onSubmit: userSubmit,
        instructionText: question.instructionText,
        isActiveQuestion: state.activeQuestion === index
      });
    } else {
      return /*#__PURE__*/_react.default.createElement(_reactNative.Text, null, "ERROR HERE");
    }
  })), /*#__PURE__*/_react.default.createElement(_buttons.ContinueButton, {
    onContinue: nextQuestion,
    labelStyle: continueLabelStyle,
    buttonStyle: continueButtonStyle,
    enabled: state.continueEnabled
  }));
};

const styles = _reactNative.StyleSheet.create({
  animatedContainer: {
    flexDirection: 'row',
    flex: 3
  }
});

var _default = QuizContainer;
exports.default = _default;
